File: problem-solving/solutions/src/test/java/blind75/TopKFrequentElementsTest.java
package blind75;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import xyz.grind.coding.blind75.TopKFrequentElements;

import static org.junit.jupiter.api.Assertions.assertAll;
import static org.junit.jupiter.api.Assertions.assertArrayEquals;

public class TopKFrequentElementsTest {

    @Test
    @DisplayName("Should return the k most frequent elements")
    void shouldReturnTrueIfFoundDuplicates() {
        assertAll(
            () -> assertArrayEquals(TopKFrequentElements.topKFrequentElementsBruteForce(new int[]{1,1,1,2,2,3,9,9,9,9,9,10,10,10,10,10,10,10,10}, 4), new int[]{10, 9, 1, 2}),
            () -> assertArrayEquals(TopKFrequentElements.topKFrequentElementsBruteForce(new int[]{34243, 98798}, 2), new int[]{34243, 98798}),
            () -> assertArrayEquals(TopKFrequentElements.topKFrequentElementsBruteForce(new int[]{1,1,1,2,3,9,11,11}, 2), new int[]{1, 11}),
            () -> assertArrayEquals(TopKFrequentElements.topKFrequentElementsHeap(new int[]{1,1,1,2,2,3,9,9,9,9,9,10,10,10,10,10,10,10,10}, 4), new int[]{10, 9, 1, 2}),
            () -> assertArrayEquals(TopKFrequentElements.topKFrequentElementsHeap(new int[]{98798,34243}, 2), new int[]{98798, 34243}),
            () -> assertArrayEquals(TopKFrequentElements.topKFrequentElementsHeap(new int[]{1,1,1,2,3,9,11,11}, 2), new int[]{1, 11}),
            () -> assertArrayEquals(TopKFrequentElements.topKFrequentElements(new int[]{1,1,1,2,3,9,11,11}, 2), new int[]{1, 11}),
            () -> assertArrayEquals(TopKFrequentElements.topKFrequentElements(new int[]{1,1,1,2,2,3,9,9,9,9,9,10,10,10,10,10,10,10,10}, 4), new int[]{10, 9, 1, 2}),
            () -> assertArrayEquals(TopKFrequentElements.topKFrequentElements(new int[]{98798,34243}, 2), new int[]{98798, 34243}),
            () -> assertArrayEquals(TopKFrequentElements.topKFrequentElements(new int[]{9876543}, 1), new int[]{9876543})
        );
    }

    @Test
    @DisplayName("Should return the k most frequent elements (edge cases)")
    void shouldReturnFalseIfNoDuplicates() {
        assertAll(
            () -> assertArrayEquals(TopKFrequentElements.topKFrequentElementsBruteForce(new int[]{10}, 1), new int[]{10}),
            () -> assertArrayEquals(TopKFrequentElements.topKFrequentElementsBruteForce(new int[]{98798,34243}, 2), new int[]{98798, 34243}),
            () -> assertArrayEquals(TopKFrequentElements.topKFrequentElementsBruteForce(new int[]{1,1,1,2,3,9,11,11}, 2), new int[]{1, 11}),
            () -> assertArrayEquals(TopKFrequentElements.topKFrequentElementsHeap(new int[]{4}, 1), new int[]{4}),
            () -> assertArrayEquals(TopKFrequentElements.topKFrequentElementsHeap(new int[]{98798,34243}, 2), new int[]{98798, 34243}),
            () -> assertArrayEquals(TopKFrequentElements.topKFrequentElementsHeap(new int[]{1,1,1,2,3,9,11,11}, 2), new int[]{1, 11}),
            () -> assertArrayEquals(TopKFrequentElements.topKFrequentElements(new int[]{11}, 1), new int[]{11}),
            () -> assertArrayEquals(TopKFrequentElements.topKFrequentElements(new int[]{1,1,1,2,2,3,9,9,9,9,9,10,10,10,10,10,10,10,10}, 4), new int[]{10, 9, 1, 2}),
            () -> assertArrayEquals(TopKFrequentElements.topKFrequentElements(new int[]{98798,34243}, 2), new int[]{98798, 34243}),
            () -> assertArrayEquals(TopKFrequentElements.topKFrequentElements(new int[]{9876543}, 1), new int[]{9876543})
        );
    }

}

File: problem-solving/solutions/leetcode/blind75/src/test/kotlin/easy/TopKFrequentElementsTest.kt
package easy

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Assertions.assertArrayEquals
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.function.Executable
import xyz.grind.coding.easy.TopKFrequentElements

class TopKFrequentElementsTest {
    @Test
    fun `Should return the k most frequent elements`() {
        Assertions.assertAll(
            Executable {
                assertArrayEquals(
                    TopKFrequentElements.topKFrequentElementsHeap(
                        intArrayOf(1, 1, 1, 2, 2, 3, 9, 9, 9, 9, 9, 10, 10, 10, 10, 10, 10, 10, 10),
                        4), intArrayOf(10, 9, 1, 2)
                )
            },
            Executable {
                assertArrayEquals(
                    TopKFrequentElements.topKFrequentElementsHeap(
                        intArrayOf(98798, 34243),
                        2
                    ), intArrayOf(98798, 34243)
                )
            },
            Executable {
                assertArrayEquals(
                    TopKFrequentElements.topKFrequentElementsHeap(
                        intArrayOf(1, 1, 1, 2, 3, 9, 11, 11),
                        2
                    ), intArrayOf(1, 11)
                )
            },
            Executable {
                assertArrayEquals(
                    TopKFrequentElements.topKFrequentElements(
                        intArrayOf(1, 1, 1, 2, 3, 9, 11, 11), 2
                    ), intArrayOf(1, 11)
                )
            },
            Executable {
                assertArrayEquals(
                    TopKFrequentElements.topKFrequentElements(
                        intArrayOf(1, 1, 1, 2, 2, 3, 9, 9, 9, 9, 9, 10, 10, 10, 10, 10, 10, 10, 10), 4),
                        intArrayOf(10, 9, 1, 2)
                )
            },
            Executable {
                assertArrayEquals(
                    TopKFrequentElements.topKFrequentElements(
                        intArrayOf(98798, 34243), 2), intArrayOf(34243, 98798)
                )
            },
            Executable {
                assertArrayEquals(
                    TopKFrequentElements.topKFrequentElements(intArrayOf(9876543), 1), intArrayOf(9876543)
                )
            }
        )
    }

    @Test
    fun `Should return the k most frequent elements (edge cases)`() {
        Assertions.assertAll(
            Executable {
                assertArrayEquals(
                    TopKFrequentElements.topKFrequentElementsHeap(intArrayOf(4), 1), intArrayOf(4))

            },
            Executable {
                assertArrayEquals(
                    TopKFrequentElements.topKFrequentElementsHeap(intArrayOf(98798, 34243), 2), intArrayOf(98798, 34243))
            },
            Executable {
                assertArrayEquals(
                    TopKFrequentElements.topKFrequentElementsHeap(
                        intArrayOf(1, 1, 1, 2, 3, 9, 11, 11), 2), intArrayOf(1, 11))
            },
            Executable {
                assertArrayEquals(
                    TopKFrequentElements.topKFrequentElements(
                        intArrayOf(11), 1), intArrayOf(11))
            },
            Executable {
                assertArrayEquals(
                    TopKFrequentElements.topKFrequentElements(
                        intArrayOf(1, 1, 1, 2, 2, 3, 9, 9, 9, 9, 9, 10, 10, 10, 10, 10, 10, 10, 10), 4
                    ), intArrayOf(10, 9, 1, 2)
                )
            },
            Executable {
                assertArrayEquals(
                    TopKFrequentElements.topKFrequentElements(intArrayOf(98798, 34243), 2), intArrayOf(34243, 98798)
                )
            },
            Executable {
                assertArrayEquals(
                    TopKFrequentElements.topKFrequentElements(intArrayOf(9876543), 1), intArrayOf(9876543))
            }
        )
    }
}
